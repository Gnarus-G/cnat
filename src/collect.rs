use std::path::PathBuf;

use swc_common::errors::{ColorConfig, Handler};
use swc_common::sync::Lrc;
use swc_common::{FileName, SourceMap};
use swc_css::visit::{Visit, VisitWith};

use swc_css::{ast::Rule, parser::parse_file};

pub struct ClassNamesCollector {
    pub class_names: Vec<cnat::Str>,
}

impl ClassNamesCollector {
    pub fn new() -> Self {
        ClassNamesCollector {
            class_names: vec![],
        }
    }

    pub fn parse(css_file: PathBuf) -> anyhow::Result<Self> {
        let code = std::fs::read_to_string(&css_file)?;

        let options = swc_css::parser::parser::ParserConfig::default();

        let cm: Lrc<SourceMap> = Default::default();
        let filename = FileName::Real(css_file);
        let cssfile = cm.new_source_file(filename.clone(), code);

        let handler = Handler::with_tty_emitter(ColorConfig::Auto, true, false, Some(cm.clone()));

        let mut errors = vec![];
        let c = parse_file::<Vec<Rule>>(&cssfile, None, options, &mut errors).unwrap();

        for e in errors {
            e.to_diagnostics(&handler).emit();
        }

        let mut ccns = ClassNamesCollector::new();

        c.visit_with(&mut ccns);

        Ok(ccns)
    }
}

impl Visit for ClassNamesCollector {
    fn visit_compound_selector(&mut self, n: &swc_css::ast::CompoundSelector) {
        let selectors = &n.subclass_selectors;

        selectors
            .iter()
            .filter_map(|s| match s {
                swc_css::ast::SubclassSelector::Class(selector) => Some(selector),
                _ => None,
            })
            .for_each(|s| {
                if s.text.value.contains(':') {
                    let cn = s.text.value.split(':').last().expect("should have at least one value after split, since empty selectors aren't allowed");
                    self.class_names.push(cn.into());
                } else {
                    self.class_names.push(s.text.value.as_str().into());
                }
            });
    }
}
