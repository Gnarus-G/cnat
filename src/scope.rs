use std::str::FromStr;

use anyhow::{anyhow, Context};
use colored::Colorize;

use crate::{Array, Str};

#[derive(Debug, PartialEq, Clone, Copy)]
pub enum ScopeVariant {
    AttrNames,
    RecordEntries,
    FnCall,
}

#[derive(Debug, Default, PartialEq, Clone, Copy)]
pub enum MatchType {
    #[default]
    Is,
    Contains,
    StartsWith,
    EndWith,
}

#[derive(Debug, PartialEq, Clone)]
pub struct ScopeValue(MatchType, Str);

#[derive(Debug, PartialEq, Clone)]
pub struct Scope {
    variant: ScopeVariant,
    values: Array<ScopeValue>,
}

impl Scope {
    pub fn matches(&self, s: &str, s_variant: ScopeVariant) -> bool {
        for value in self.values.iter() {
            let matches = match value.0 {
                MatchType::Contains => s.contains(&*value.1),
                MatchType::Is => *value.1 == *s,
                MatchType::StartsWith => s.starts_with(&*value.1),
                MatchType::EndWith => s.ends_with(&*value.1),
            };

            if matches && self.variant == s_variant {
                return true;
            }
        }

        false
    }
}

impl FromStr for Scope {
    type Err = anyhow::Error;

    ///grammar -> variant:value,value,*value,...,value
    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let group = s.split(':').collect::<Vec<_>>();

        let [variant, values] = group.as_slice() else {
            return Err(anyhow!("incorrect number of parts: {:?}", group))
                .with_context(|| {
                    return format!("\n{}\n\tvariants are one of {}\n\ta value can be a string starting or ending with an '*'", "should be two parts, a variant and values: <variant>:<...values>".yellow(), "att | prop | fn".green())
                });
        };

        let values = values
            .split(',')
            .filter(|v| !v.is_empty())
            .map(|v| {
                let mut mt = MatchType::default();

                let [identifier] = v.split('*').filter(|v| !v.is_empty()).collect::<Vec<_>>()[..]
                else {
                    return Err(anyhow!("cannot have a wildcard in the middle"));
                };

                if v.starts_with('*') && v.ends_with('*') {
                    mt = MatchType::Contains;
                } else if v.starts_with('*') {
                    mt = MatchType::EndWith;
                } else if v.ends_with('*') {
                    mt = MatchType::StartsWith;
                }

                Ok(ScopeValue(mt, identifier.into()))
            })
            .collect::<Result<Vec<_>, anyhow::Error>>()?;

        if values.is_empty() {
            return Err(anyhow!("at least one value must be provided"));
        }

        let values = values.into();

        let variant = match *variant {
            "att" => ScopeVariant::AttrNames,
            "prop" => ScopeVariant::RecordEntries,
            "fn" => ScopeVariant::FnCall,
            _ => return Err(anyhow!("unrecognized variant: {}", variant)),
        };

        Ok(Scope { variant, values })
    }
}

#[cfg(test)]
mod tests {
    use std::str::FromStr;

    use crate::scope::{MatchType, Scope, ScopeValue, ScopeVariant};

    #[test]
    fn it_parses() {
        assert_eq!(
            Scope::from_str("att:class,className,iconClassName").unwrap(),
            Scope {
                variant: ScopeVariant::AttrNames,
                values: vec![
                    ScopeValue(MatchType::Is, "class".into()),
                    ScopeValue(MatchType::Is, "className".into()),
                    ScopeValue(MatchType::Is, "iconClassName".into())
                ]
                .into_boxed_slice()
            }
        );

        assert_eq!(
            Scope::from_str("prop:classes,className,").unwrap(),
            Scope {
                variant: ScopeVariant::RecordEntries,
                values: vec![
                    ScopeValue(MatchType::Is, "classes".into(),),
                    ScopeValue(MatchType::Is, "className".into())
                ]
                .into_boxed_slice()
            }
        );

        assert_eq!(
            Scope::from_str("fn:cva").unwrap(),
            Scope {
                variant: ScopeVariant::FnCall,
                values: vec![ScopeValue(MatchType::Is, "cva".into())].into_boxed_slice()
            }
        )
    }

    #[test]
    fn it_parses_starting_wildcard() {
        assert_eq!(
            Scope::from_str("att:class,*ClassName").unwrap(),
            Scope {
                variant: ScopeVariant::AttrNames,
                values: vec![
                    ScopeValue(MatchType::Is, "class".into()),
                    ScopeValue(MatchType::EndWith, "ClassName".into()),
                ]
                .into_boxed_slice()
            }
        );

        assert_eq!(
            Scope::from_str("prop:classes,***ClassName,").unwrap(),
            Scope {
                variant: ScopeVariant::RecordEntries,
                values: vec![
                    ScopeValue(MatchType::Is, "classes".into(),),
                    ScopeValue(MatchType::EndWith, "ClassName".into())
                ]
                .into_boxed_slice()
            }
        );
    }

    #[test]
    fn it_parses_ending_wildcard() {
        assert_eq!(
            Scope::from_str("att:class,class**").unwrap(),
            Scope {
                variant: ScopeVariant::AttrNames,
                values: vec![
                    ScopeValue(MatchType::Is, "class".into()),
                    ScopeValue(MatchType::StartsWith, "class".into()),
                ]
                .into_boxed_slice()
            }
        );

        assert_eq!(
            Scope::from_str("prop:class**").unwrap(),
            Scope {
                variant: ScopeVariant::RecordEntries,
                values: vec![ScopeValue(MatchType::StartsWith, "class".into())].into_boxed_slice()
            }
        );
    }

    #[test]
    fn it_rejects_middle_wildcard() {
        Scope::from_str("att:class,class*name").unwrap_err();
        Scope::from_str("prop:class*name").unwrap_err();
    }

    #[test]
    fn it_rejects_empty_values() {
        Scope::from_str("att:").unwrap_err();
        Scope::from_str("prop:,").unwrap_err();
    }

    #[test]
    fn it_matches() {
        let scope = Scope {
            variant: ScopeVariant::RecordEntries,
            values: vec![
                ScopeValue(MatchType::Is, "classes".into()),
                ScopeValue(MatchType::Is, "className".into()),
            ]
            .into_boxed_slice(),
        };

        assert!(!scope.matches("className", ScopeVariant::AttrNames));
        assert!(!scope.matches("class", ScopeVariant::RecordEntries));

        assert!(scope.matches("className", ScopeVariant::RecordEntries));
    }

    #[test]
    fn it_matches_ends() {
        let scope = Scope {
            variant: ScopeVariant::AttrNames,
            values: vec![ScopeValue(MatchType::EndWith, "ClassName".into())].into_boxed_slice(),
        };

        assert!(!scope.matches("className", ScopeVariant::AttrNames));
        assert!(!scope.matches("class", ScopeVariant::RecordEntries));

        assert!(scope.matches("iconClassName", ScopeVariant::AttrNames));
        assert!(scope.matches("bodyClassName", ScopeVariant::AttrNames));
        assert!(scope.matches("buttonClassName", ScopeVariant::AttrNames));
    }

    #[test]
    fn it_matches_starts() {
        let scope = Scope {
            variant: ScopeVariant::AttrNames,
            values: vec![ScopeValue(MatchType::StartsWith, "class".into())].into_boxed_slice(),
        };

        assert!(!scope.matches("class", ScopeVariant::RecordEntries));
        assert!(scope.matches("class", ScopeVariant::AttrNames));
        assert!(scope.matches("className", ScopeVariant::AttrNames));
    }

    #[test]
    fn it_matches_contains() {
        let scope = Scope {
            variant: ScopeVariant::AttrNames,
            values: vec![ScopeValue(MatchType::Contains, "class".into())].into_boxed_slice(),
        };

        assert!(!scope.matches("class", ScopeVariant::RecordEntries));
        assert!(scope.matches("class", ScopeVariant::AttrNames));
        assert!(scope.matches("className", ScopeVariant::AttrNames));
        assert!(scope.matches("firstclassName", ScopeVariant::AttrNames));
        assert!(scope.matches("buttonclassName", ScopeVariant::AttrNames));

        let scope = Scope {
            variant: ScopeVariant::AttrNames,
            values: vec![ScopeValue(MatchType::Contains, "Class".into())].into_boxed_slice(),
        };

        assert!(scope.matches("iconClassName", ScopeVariant::AttrNames));
        assert!(scope.matches("bodyClassName", ScopeVariant::AttrNames));
        assert!(scope.matches("buttonClassName", ScopeVariant::AttrNames));
    }
}
