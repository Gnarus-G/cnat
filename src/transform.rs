use anyhow::{anyhow, Context};
use colored::Colorize;
use std::ffi::OsStr;
use std::path::Path;
use std::vec;
use swc_common::sync::Lrc;
use swc_common::{
    errors::{ColorConfig, Handler},
    SourceMap,
};
use swc_ecma_ast::{Callee, EsVersion, Expr, Ident, JSXAttrName, PropName};
use swc_ecma_parser::{parse_file_as_program, Syntax};
use swc_ecma_visit::{VisitMut, VisitMutWith};

use cnat::scope::{Scope, ScopeVariant};

pub struct ApplyTailwindPrefix<'s, 'cn, 'scopes> {
    pub prefix: &'s str,
    class_names: &'cn [cnat::Str],
    scopes: &'scopes [Scope],
    is_in_scope: bool,
    replacements: Vec<replacements::Replacement>,
}

impl<'s, 'cn, 'scopes> ApplyTailwindPrefix<'s, 'cn, 'scopes> {
    pub fn new(prefix: &'s str, class_names: &'cn [cnat::Str], scopes: &'scopes [Scope]) -> Self {
        Self {
            prefix,
            class_names,
            scopes,
            is_in_scope: false,
            replacements: vec![],
        }
    }

    /// Returns the number of files transformed.
    pub fn prefix_all_classes_in_dir(&mut self, path: &Path) -> anyhow::Result<usize> {
        assert!(path.is_dir());

        let mut edit_count = 0;

        for r in ignore::Walk::new(path) {
            match r {
                Ok(entry) => {
                    let filepath = entry.path();
                    let is_supported_file = filepath.is_file()
                        && filepath
                            .extension()
                            .map(|e| ["ts", "js", "jsx", "tsx"].map(OsStr::new).contains(&e))
                            .unwrap_or(false);

                    if !is_supported_file {
                        continue;
                    }

                    match self.prefix_classes_in_file(filepath) {
                        Ok(Some(())) => {
                            edit_count += 1;
                        }
                        Err(err) => {
                            eprintln!(
                                "{} failed to process file, {}: {err:#}",
                                "[ERROR]".red(),
                                filepath.display()
                            )
                        }
                        Ok(None) => {}
                    }
                }
                Err(err) => eprintln!("[Error] {err:#}"),
            };
        }

        Ok(edit_count)
    }

    pub fn prefix_classes_in_file(&mut self, source_file: &Path) -> anyhow::Result<Option<()>> {
        let cm: Lrc<SourceMap> = Default::default();
        let error_handler =
            Handler::with_tty_emitter(ColorConfig::Auto, true, false, Some(cm.clone()));

        let fm = cm
            .load_file(source_file)
            .context("failed to load source file")?;

        let syntax = match source_file.extension().and_then(|e| e.to_str()) {
            Some("js") | Some("jsx") => Syntax::Es(swc_ecma_parser::EsConfig {
                jsx: true,
                ..Default::default()
            }),
            Some("ts") => Syntax::Typescript(Default::default()),
            Some("tsx") => Syntax::Typescript(swc_ecma_parser::TsConfig {
                tsx: true,
                ..Default::default()
            }),
            None => {
                return Err(anyhow!(
                    "unknown filetype, missing extension: {}",
                    source_file.display()
                ))
            }
            ext => return Err(anyhow!("unknown filetype: {ext:?}")),
        };

        let mut errors = vec![];
        let mut program = parse_file_as_program(&fm, syntax, EsVersion::Es2015, None, &mut errors)
            .map_err(|e| e.into_diagnostic(&error_handler).emit())
            .expect("failed to parse source code file");

        program.visit_mut_children_with(self);

        if self.replacements.is_empty() {
            return Ok(None);
        }

        let contents = std::fs::read(source_file).context("failed to file for writing")?;

        eprintln!("[INFO] reading to transform {}", source_file.display());

        let contents = replacements::Replacement::apply_all(&mut self.replacements, contents);
        std::fs::write(source_file, contents)?;

        eprintln!(
            "[INFO] transformed {}",
            source_file.display().to_string().green()
        );

        self.replacements.clear();

        Ok(Some(()))
    }

    fn starts_a_valid_scope(&self, ident: &Ident, variant: ScopeVariant) -> bool {
        let ident = ident.sym.as_str();
        self.scopes
            .iter()
            .any(|scope| scope.matches(ident, variant))
    }
}

impl<'s, 'cn, 'scopes> VisitMut for ApplyTailwindPrefix<'s, 'cn, 'scopes> {
    fn visit_mut_jsx_attr(&mut self, n: &mut swc_ecma_ast::JSXAttr) {
        if let JSXAttrName::Ident(name) = &n.name {
            if self.starts_a_valid_scope(name, ScopeVariant::AttrNames) {
                self.is_in_scope = true;
                n.value.visit_mut_with(self);
                self.is_in_scope = false;
            }
        }

        n.visit_mut_children_with(self);
    }

    fn visit_mut_call_expr(&mut self, n: &mut swc_ecma_ast::CallExpr) {
        if let Callee::Expr(expr) = &n.callee {
            if let Expr::Ident(name) = expr.as_ref() {
                if self.starts_a_valid_scope(name, ScopeVariant::FnCall) {
                    self.is_in_scope = true;
                    n.args.visit_mut_with(self);
                    self.is_in_scope = false;
                }
            }
        }

        n.visit_mut_children_with(self);
    }

    fn visit_mut_key_value_prop(&mut self, n: &mut swc_ecma_ast::KeyValueProp) {
        if let PropName::Ident(ident) = &n.key {
            if self.starts_a_valid_scope(ident, ScopeVariant::RecordEntries) {
                self.is_in_scope = true;
                n.value.visit_mut_with(self);
                self.is_in_scope = false;
            }
        }

        n.visit_mut_children_with(self);
    }

    fn visit_mut_str(&mut self, n: &mut swc_ecma_ast::Str) {
        if !self.is_in_scope {
            return;
        }

        let mut has_prefixed_some = false;
        let replacements: Vec<_> = n
            .value
            .split(' ')
            .map(|class| {
                if class.is_empty() {
                    return class.to_string();
                }

                let mut class_fragments: Vec<_> = class.split(':').collect();
                let actual_class = class_fragments
                    .last_mut()
                    .expect("class should not have been an empty string");

                if self.class_names.iter().any(|name| name == *actual_class) {
                    let prefixed = format!("{}{}", self.prefix, actual_class);
                    *actual_class = prefixed.as_str();
                    has_prefixed_some = true;
                    return class_fragments.join(":");
                }

                class.to_string()
            })
            .collect();

        if has_prefixed_some {
            let start = n.span.lo.0 as usize - 1; // - 1 because swc bytepos is 1-based
            let end = n.span.hi.0 as usize - 1;

            // exclude the begining and end quotes counted in the span
            let start = start + 1;
            let end = end - 2;

            debug_assert_eq!(
                end - start + 1, // computed value length
                n.value.as_bytes().len()
            );

            let replacement = replacements.join(" ");

            self.replacements.push(replacements::Replacement::new(
                start..=end,
                n.value.as_bytes(),
                replacement.as_bytes(),
            ));
        }
    }
}

mod replacements {

    pub struct Replacement {
        byte_range: std::ops::RangeInclusive<usize>,
        old: cnat::Array<u8>,
        new: cnat::Array<u8>,
    }

    impl Replacement {
        pub fn new(byte_range: std::ops::RangeInclusive<usize>, old: &[u8], new: &[u8]) -> Self {
            Self {
                byte_range,
                old: old.into(),
                new: new.into(),
            }
        }

        fn slide_span(&mut self, addition: usize) {
            let start = self.byte_range.start() + addition;
            let end = self.byte_range.end() + addition;
            self.byte_range = start..=end;
        }

        fn apply(&mut self, contents: &mut Vec<u8>, byte_additions: usize) -> usize {
            self.slide_span(byte_additions);

            let to_be_removed = &contents[self.byte_range.clone()];
            debug_assert_eq!(
                String::from_utf8_lossy(self.old.as_ref()),
                String::from_utf8_lossy(to_be_removed)
            );
            assert_eq!(
                self.old.as_ref(),
                to_be_removed,
                "invariant failed: the range to replace is not equal to string value parsed"
            );

            let replace_with = self.new.iter().cloned();
            contents.splice(self.byte_range.clone(), replace_with);

            let addition = self.new.len().saturating_sub(self.old.len());
            return addition;
        }

        pub fn apply_all(rps: &mut [Replacement], mut contents: Vec<u8>) -> Vec<u8> {
            let mut byte_additions = 0;
            for rp in rps {
                byte_additions += rp.apply(&mut contents, byte_additions);
            }
            return contents;
        }
    }

    #[cfg(test)]
    mod tests {
        use super::*;

        #[test]
        fn replacements() {
            let contents = "1234567hiearth".as_bytes().to_vec();
            let rps = &mut [
                Replacement::new(1..=3, "234".as_bytes(), "abcdef".as_bytes()),
                Replacement::new(5..=6, "67".as_bytes(), "jkl".as_bytes()),
                Replacement::new(7..=8, "hi".as_bytes(), "hello".as_bytes()),
                Replacement::new(9..=13, "earth".as_bytes(), "world".as_bytes()),
            ];

            let contents = Replacement::apply_all(rps, contents);

            assert_eq!(contents, "1abcdef5jklhelloworld".as_bytes());
        }
    }
}
