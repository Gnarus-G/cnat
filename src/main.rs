mod collect;
mod transform;

use std::path::PathBuf;

use anyhow::anyhow;
use clap::{crate_name, Args, CommandFactory, Parser, Subcommand, ValueHint};
use cnat::scope::Scope;
use collect::ClassNamesCollector;
use colored::Colorize;

use crate::transform::ApplyTailwindPrefix;

/// Systematically apply certain modifications to classes, class names, used
/// in your frontend codebase.
#[derive(Parser)]
#[clap(about, author, version)]
struct Cli {
    #[command(subcommand)]
    command: Command,
}

#[derive(Subcommand)]
enum Command {
    /// Apply a prefix to all the tailwind classes in every js file in a project.
    Prefix(PrefixArgs),

    /// Generate completions for a specified shell
    Completion {
        // The shell for which to generate completions
        shell: clap_complete::Shell,
    },
}

#[derive(Args)]
struct PrefixArgs {
    /// The output css file generated by calling `npx tailwindcss -i input.css -o output.css`
    #[arg(short = 'i', value_hint = ValueHint::FilePath)]
    css_file: PathBuf,

    /// The prefix to apply to all the tailwind class names found
    #[arg(short, long)]
    prefix: String,

    /// Define scope within which prefixing happens. Example: --scopes 'att:className,*ClassName prop:classes fn:cva'
    #[arg(short, long, num_args = 1.., value_delimiter = ' ', default_value = "att:class,className fn:createElement")]
    scopes: Vec<Scope>,

    /// The directories in which to find js/ts files.
    #[arg(value_hint = ValueHint::DirPath)]
    contexts: Vec<PathBuf>,
}

fn main() -> anyhow::Result<()> {
    let cli = Cli::parse();

    let cli = match cli.command {
        Command::Prefix(cli) => cli,
        Command::Completion { shell } => {
            clap_complete::generate(
                shell,
                &mut Cli::command(),
                crate_name!(),
                &mut std::io::stdout(),
            );
            return Ok(());
        }
    };

    for context in &cli.contexts {
        if !context.is_dir() {
            return Err(anyhow!(
                "context should be a directory, got {}",
                context.display()
            ));
        }
    }

    let c = ClassNamesCollector::parse(cli.css_file)?;

    eprintln!("[INFO] extracted selectors");
    println!("{:?}", c.class_names);

    let mut ppc = ApplyTailwindPrefix::new(&cli.prefix, &c.class_names, &cli.scopes);

    for context in &cli.contexts {
        ppc.prefix_all_classes_in_dir(context)?;
    }

    eprintln!("{}", "[DONE] Remember to run your formatter on the transformed files to make sure the format is as expected.".green());

    Ok(())
}

#[cfg(test)]
mod tests {
    use assert_cmd::Command;
    use insta::assert_snapshot;
    use std::{fs, path::PathBuf};

    struct JsFile(PathBuf, Vec<u8>);

    impl JsFile {
        fn prep(path: &'static str, temp_dir: &str) -> Self {
            let js_file_content_before = fs::read(path).expect("failed to read js fixture file");

            let new_path = std::path::PathBuf::from(format!("{}/{}", temp_dir, path));
            fs::create_dir_all(new_path.parent().unwrap()).unwrap();
            fs::copy(path, &new_path).unwrap();

            Self(new_path, js_file_content_before)
        }

        fn content_now(&self) -> String {
            let js_file_content_after = fs::read(&self.0).expect("failed to read js fixture file");
            String::from_utf8_lossy(&js_file_content_after).to_string()
        }
    }

    impl Drop for JsFile {
        fn drop(&mut self) {
            fs::remove_file(&self.0).expect("failed to remove a file")
        }
    }

    #[test]
    fn it_works_with_default_scopes() {
        let context_dir = "basic";
        let jsfiles = [
            JsFile::prep("fixtures/sample.tsx", context_dir),
            JsFile::prep("fixtures/nested/sample.tsx", context_dir),
            JsFile::prep("fixtures/nested/nested/sample.tsx", context_dir),
            JsFile::prep("fixtures/sample2.tsx", context_dir),
            JsFile::prep("fixtures/nested/sample2.tsx", context_dir),
        ];

        let cssfile = "fixtures/sample.css";
        let mut cmd = Command::cargo_bin("cnat").unwrap();
        let cmd = cmd
            .args(["prefix", "-i", cssfile, "--prefix", "tw-", context_dir])
            .assert()
            .success();

        let output = cmd.get_output();

        let output = String::from_utf8_lossy(&output.stdout);

        insta::with_settings!({
            info => &cssfile,
            omit_expression => true
        }, {
            assert_snapshot!(output);
        });

        for jsfile in jsfiles {
            insta::with_settings!({
                snapshot_suffix => jsfile.0.to_string_lossy(),
                info => &jsfile.0,
                description => output.clone(),
                omit_expression => true
            }, {
                assert_snapshot!(jsfile.content_now());
            });
        }
    }

    #[test]
    fn it_works_with_cva_fn_scope() {
        let context_dir = "cva";
        let jsfiles = [
            JsFile::prep("fixtures/nested/sample.tsx", context_dir),
            JsFile::prep("fixtures/sample2.tsx", context_dir),
        ];

        let cssfile = "fixtures/sample.css";
        let scopes = "fn:cva";
        let mut cmd = Command::cargo_bin("cnat").unwrap();
        let cmd = cmd
            .args([
                "prefix",
                "-i",
                cssfile,
                "--prefix",
                "tw-",
                context_dir,
                "--scopes",
                scopes,
            ])
            .assert()
            .success();

        let output = cmd.get_output();

        let output = String::from_utf8_lossy(&output.stdout);

        insta::with_settings!({
            info => &cssfile,
            omit_expression => true
        }, {
            assert_snapshot!(output);
        });

        for jsfile in jsfiles {
            insta::with_settings!({
                snapshot_suffix => jsfile.0.to_string_lossy(),
                info => &jsfile.0,
                description => scopes,
                omit_expression => true
            }, {
                assert_snapshot!(jsfile.content_now());
            });
        }
    }

    #[test]
    fn it_works_with_custom_jsx_attribute() {
        let context_dir = "object_inside";
        let jsfiles = [
            JsFile::prep("fixtures/nested/sample.tsx", context_dir),
            JsFile::prep("fixtures/nested/nested/sample.tsx", context_dir),
            JsFile::prep("fixtures/sample2.tsx", context_dir),
            JsFile::prep("fixtures/nested/sample2.tsx", context_dir),
        ];

        let cssfile = "fixtures/sample.css";
        let scopes = "att:classes,*ClassName";
        let mut cmd = Command::cargo_bin("cnat").unwrap();
        let cmd = cmd
            .args([
                "prefix",
                "-i",
                cssfile,
                "--prefix",
                "tw-",
                context_dir,
                "--scopes",
                scopes,
            ])
            .assert()
            .success();

        let output = cmd.get_output();

        let output = String::from_utf8_lossy(&output.stdout);

        insta::with_settings!({
            info => &cssfile,
            omit_expression => true
        }, {
            assert_snapshot!(output);
        });

        for jsfile in jsfiles {
            insta::with_settings!({
                snapshot_suffix => jsfile.0.to_string_lossy(),
                info => &jsfile.0,
                description => scopes,
                omit_expression => true
            }, {
                assert_snapshot!(jsfile.content_now());
            });
        }
    }

    #[test]
    fn it_works_with_classes_or_classname_object_entries() {
        let context_dir = "object_outside";
        let jsfiles = [
            JsFile::prep("fixtures/sample.tsx", context_dir),
            JsFile::prep("fixtures/nested/sample.tsx", context_dir),
            JsFile::prep("fixtures/nested/nested/sample.tsx", context_dir),
            JsFile::prep("fixtures/sample2.tsx", context_dir),
            JsFile::prep("fixtures/nested/sample2.tsx", context_dir),
        ];

        let cssfile = "fixtures/sample.css";
        let scopes = "prop:classes prop:className";
        let mut cmd = Command::cargo_bin("cnat").unwrap();
        let cmd = cmd
            .args([
                "prefix",
                "-i",
                cssfile,
                "--prefix",
                "tw-",
                context_dir,
                "--scopes",
                scopes,
            ])
            .assert()
            .success();

        let output = cmd.get_output();

        let output = String::from_utf8_lossy(&output.stdout);

        insta::with_settings!({
            info => &cssfile,
            omit_expression => true
        }, {
            assert_snapshot!(output);
        });

        for jsfile in jsfiles {
            insta::with_settings!({
                snapshot_suffix => jsfile.0.to_string_lossy(),
                info => &jsfile.0,
                description => scopes,
                omit_expression => true
            }, {
                assert_snapshot!(jsfile.content_now());
            });
        }
    }

    #[test]
    fn it_preserves_comments() {
        let context_dir = "preserves_comments";
        let jsfile = JsFile::prep("fixtures/sample_comments.jsx", context_dir);

        let cssfile = "fixtures/sample.css";
        let mut cmd = Command::cargo_bin("cnat").unwrap();
        let cmd = cmd
            .args(["prefix", "-i", cssfile, "--prefix", "tw-", context_dir])
            .assert()
            .success();

        let output = cmd.get_output();

        let output = String::from_utf8_lossy(&output.stdout);

        insta::with_settings!({
            info => &cssfile,
            omit_expression => true
        }, {
            assert_snapshot!(output);
        });

        insta::with_settings!({
            snapshot_suffix => jsfile.0.to_string_lossy(),
            info => &jsfile.0,
            omit_expression => true
        }, {
            assert_snapshot!(jsfile.content_now());
        });
    }

    #[test]
    fn it_leaves_alone_files_without_classes_to_prefix() {
        let context_dir = "leave_unstyled";
        let jsfile = JsFile::prep("fixtures/unstyled.tsx", context_dir);

        let cssfile = "fixtures/sample.css";
        let mut cmd = Command::cargo_bin("cnat").unwrap();
        let cmd = cmd
            .args(["prefix", "-i", cssfile, "--prefix", "tw-", context_dir])
            .assert()
            .success();

        let output = cmd.get_output();

        let output = String::from_utf8_lossy(&output.stdout);

        insta::with_settings!({
            info => &cssfile,
            omit_expression => true
        }, {
            assert_snapshot!(output);
        });

        insta::with_settings!({
            snapshot_suffix => jsfile.0.to_string_lossy(),
            info => &jsfile.0,
            omit_expression => true
        }, {
            assert_snapshot!(jsfile.content_now());
        });
    }
}
