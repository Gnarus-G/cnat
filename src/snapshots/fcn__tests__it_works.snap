---
source: src/main.rs
info: fixtures/sample.css
---
["sr-only", "!visible", "visible", "mx-1", "w-[10px]", "absolute", "MuiButton-startIcon", "pb-6", "text-sm", "w-10/11", "w-10"]
